OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,432 @@
+			var reader = new BinaryReader(fil);
+			if (reader.ReadUInt32() != 0x28635349)
+				throw new InvalidDataException("Not an Installshield CAB package");
+			reader.BaseStream.Seek(fd.DataOffset, SeekOrigin.Begin);
+			var destfile = File.Open(fileName, FileMode.Create);
+			var writer = new BinaryWriter(destfile);
+			if ((fd.Flags & FILECOMPRESSED) != 0)
+			{
+				uint bytes_to_read = fd.CompressedSize;
+				ushort bytes_to_extract;
+				byte[] read_buffer;
+				const int BUFFER_SIZE = 65536;
+				var write_buffer = new byte[BUFFER_SIZE];
+				int extracted_bytes;
+				var inf = new ICSharpCode.SharpZipLib.Zip.Compression.Inflater(true);
+				while (bytes_to_read > 0)
+				{
+					bytes_to_extract = reader.ReadUInt16();
+					read_buffer = reader.ReadBytes(bytes_to_extract);
+					inf.SetInput(read_buffer);
+					extracted_bytes = inf.Inflate(write_buffer);
+					if (extracted_bytes > BUFFER_SIZE)
+						throw new Exception("needs a bigger buffer");
+					writer.Write(write_buffer, 0, extracted_bytes);
+					bytes_to_read -= (uint)bytes_to_extract + 2;
+
+					inf.Reset();
+				}
+
+				writer.Dispose();
+			}
+			else
+			{
+				writer.Write(reader.ReadBytes((int)fd.ExpandedSize));
+				writer.Dispose();
+			}
+
+			fil.Dispose();
+		}
+
+		public void Write(Dictionary<string, byte[]> input)
+		{
+			throw new NotImplementedException("Cannot Add Files To Cab");
+		}
+
+		public IEnumerable<uint> ClassicHashes()
+		{
+			return fileDict.Keys.Select(k => PackageEntry.HashFilename(k, PackageHashType.Classic));
+		}
+
+		public Stream GetContent(string fileName)
+		{
+			var file_des = fileDict[fileName];
+			if ((file_des.Flags & FILEINVALID) != 0) throw new Exception("File Invalid");
+			if ((file_des.LinkFlags & LINKPREV) != 0)
+				throw new NotImplementedException("Link Previous");
+
+			if ((file_des.Flags & FILESPLIT) != 0 || (file_des.Flags & FILEOBFUSCATED) != 0)
+				throw new NotImplementedException("Haven't implemented split or obfustcated files");
+
+			var out_stream = new MemoryStream();
+			using (var file_stream = GlobalFileSystem.Open(string.Format("{0}{1}.cab", commonName, file_des.Volume)))
+			{
+				var reader = new BinaryReader(file_stream);
+				if (reader.ReadUInt32() != 0x28635349)
+					throw new InvalidDataException("Not an Installshield CAB package");
+				file_stream.Seek(file_des.DataOffset, SeekOrigin.Begin);
+				if ((file_des.Flags & FILECOMPRESSED) != 0)
+				{
+					uint bytes_to_read = file_des.CompressedSize;
+					ushort bytes_to_extract;
+					var buffer = new byte[4096];
+					var inflate = new InflaterInputStream(file_stream);
+					while (bytes_to_read > 0)
+					{
+						bytes_to_extract = reader.ReadUInt16();
+						while (bytes_to_extract > 0)
+						{
+							var len = inflate.Read(buffer, 0, bytes_to_extract);
+							bytes_to_extract -= (ushort)len;
+							out_stream.Write(buffer, 0, len);
+						}
+
+						bytes_to_read -= bytes_to_extract;
```

Comment:
If you convert all those underscored variables to camelCase this will look a lot nicer and adhere to our https://github.com/OpenRA/OpenRA/wiki/Coding-Standard